plugins/microservices/src/test_msp_functions.cpp

```diff
@@ -0,0 +1,94 @@
+/// \file
+
+#include "irods/irods_ms_plugin.hpp"
+#include "irods/msParam.h"
+
+#include "msi_assertion_macros.hpp"
+
+namespace
+{
+    auto test_null_cases([[maybe_unused]] RuleExecInfo* _rei) -> int
+    {
+        IRODS_MSI_TEST_BEGIN("testing msp_to_string: msp_to_string null cases")
+
+        char* good = nullptr;
+        msParam_t in{};
+
+        auto ret = msp_to_string(nullptr, nullptr);
+        IRODS_MSI_ASSERT(SYS_NULL_INPUT == ret);
+
+        ret = msp_to_string(nullptr, &good);
+        IRODS_MSI_ASSERT((SYS_NULL_INPUT == ret && nullptr == good));
```

Comment:
Consider separating individual checks into multiple assertions. That'll make it easy to see which one causes a failure.

_Same thing for other assertions._